r1cs/gadgets/crypto/src/prf/ripemd160.rs

```diff
@@ -0,0 +1,458 @@
+        _ => unreachable!()
+    };
+    (UInt32::constant(k), UInt32::constant(k_prime))
+}
+
+#[allow(clippy::unreadable_literal)]
+const IV: [u32; 5] = [
+    0x67452301, 0xefcdab89, 0x98badcfe, 0x10325476, 0xc3d2e1f0
+];
+
+/// Amount for first rotate left
+const S: [usize; 80] = [
+    11, 14, 15, 12, 5, 8, 7, 9, 11, 13, 14, 15, 6, 7, 9, 8,
+    7, 6, 8, 13, 11, 9, 7, 15, 7, 12, 15, 9, 11, 7, 13, 12,
+    11, 13, 6, 7, 14, 9, 13, 15, 14, 8, 13, 6, 5, 12, 7, 5,
+    11, 12, 14, 15, 14, 15, 9, 8, 9, 14, 5, 6, 8, 6, 5, 12,
+    9, 15, 5, 11, 6, 8, 13, 12, 5, 12, 13, 14, 11, 8, 5, 6
+];
+
+/// Amount for second rotate left
+const S_PRIME: [usize; 80] = [
+    8, 9, 9, 11, 13, 15, 15, 5, 7, 7, 8, 11, 14, 14, 12, 6,
+    9, 13, 15, 7, 12, 8, 9, 11, 7, 7, 12, 7, 6, 15, 13, 11,
+    9, 7, 15, 11, 8, 6, 6, 14, 12, 13, 5, 14, 13, 13, 7, 5,
+    15, 5, 8, 11, 14, 14, 6, 14, 6, 9, 12, 9, 12, 5, 15, 8,
+    8, 5, 12, 9, 12, 5, 14, 6, 8, 13, 6, 5, 15, 13, 11, 11
+];
+
+/// First message word selection
+const R: [usize; 80] = [
+    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
+    7, 4, 13, 1, 10, 6, 15, 3, 12, 0, 9, 5, 2, 14, 11, 8,
+    3, 10, 14, 4, 9, 15, 8, 1, 2, 7, 0, 6, 13, 11, 5, 12,
+    1, 9, 11, 10, 0, 8, 12, 4, 13, 3, 7, 15, 14, 5, 6, 2,
+    4, 0, 5, 9, 7, 12, 2, 10, 14, 1, 3, 8, 11, 6, 15, 13
+];
+
+/// Second message word selection
+const R_PRIME: [usize; 80] = [
+    5, 14, 7, 0, 9, 2, 11, 4, 13, 6, 15, 8, 1, 10, 3, 12,
+    6, 11, 3, 7, 0, 13, 5, 10, 14, 15, 8, 12, 4, 9, 1, 2,
+    15, 5, 1, 3, 7, 14, 6, 9, 11, 8, 12, 2, 10, 0, 4, 13,
+    8, 6, 4, 1, 3, 11, 15, 0, 5, 12, 2, 13, 9, 7, 10, 14,
+    12, 15, 10, 4, 1, 5, 8, 7, 6, 2, 13, 14, 0, 3, 9, 11
+];
+
+/// Apply the proper round function, selected with round_idx, to inputs a, b and c
+fn apply_round_function<ConstraintF, CS>(
+    cs:        CS,
+    a:         &UInt32,
+    b:         &UInt32,
+    c:         &UInt32,
+    round_idx: usize,
+) -> Result<UInt32, SynthesisError>
+    where
+        ConstraintF: PrimeField,
+        CS: ConstraintSystem<ConstraintF>
+{
+    // Select and return the proper round function according to round idx
+    let (tri_fn, circ_fn): (fn(u32, u32, u32) -> u32, fn(&mut CS, usize, &Boolean, &Boolean, &Boolean) -> Result<Boolean, SynthesisError>) = match round_idx {
+        // f(j, a, b, c) = a XOR b XOR c (0 <= j <= 15)
+        round_idx if round_idx <= 15 => {
+            (
+                |a, b, c| a ^ b ^ c,
+                |cs, i, a, b, c| {
+                    let result = Boolean::xor(cs.ns(|| format!("A XOR B {}", i)), &a, &b)?;
+                    Boolean::xor(cs.ns(|| format!("A XOR B XOR C {}", i)), &result,&c)
+                }
+            )
+        },
+        // f(j, a, b, c) = (a AND b) OR (NOT(a) AND c)  (16 <= j <= 31)
+        // Note: It's the same as the sha256_ch function e.g. (a AND b) XOR (NOT(a) AND c)
+        // since the two logical expressions have the same truth table
```

Comment:
Correct. Clearly the two "addends" can't be both 1, so OR and XOR acts equivalently.